File: Projects/FinSeAPDF/PDFBox/pdmodel/font/PDFontDescriptorAFM.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.pdfbox.pdmodel.font;

import java.io.IOException;

import org.apache.fontbox.afm.FontMetric;

import org.apache.pdfbox.pdmodel.common.PDRectangle;

import org.apache.fontbox.util.BoundingBox;

/**
 * This class represents the font descriptor when the font information
 * is coming from an AFM file.
 *
 * @author <a href="mailto:ben@benlitchfield.com">Ben Litchfield</a>
 * @version $Revision: 1.3 $
 */
public class PDFontDescriptorAFM extends PDFontDescriptor
{
    private FontMetric afm;

    /**
     * Constructor.
     *
     * @param afmFile The AFM file.
     */
    public PDFontDescriptorAFM( FontMetric afmFile )
    {
        afm = afmFile;
    }

    /**
     * Get the font name.
     *
     * @return The name of the font.
     */
    public String getFontName()
    {
        return afm.getFontName();
    }

    /**
     * This will set the font name.
     *
     * @param fontName The new name for the font.
     */
    public void setFontName( String fontName )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * A string representing the preferred font family.
     *
     * @return The font family.
     */
    public String getFontFamily()
    {
        return afm.getFamilyName();
    }

    /**
     * This will set the font family.
     *
     * @param fontFamily The font family.
     */
    public void setFontFamily( String fontFamily )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * The weight of the font.  According to the PDF spec "possible values are
     * 100, 200, 300, 400, 500, 600, 700, 800 or 900"  Where a higher number is
     * more weight and appears to be more bold.
     *
     * @return The font weight.
     */
    public Single getFontWeight()
    {
        String weight = afm.getWeight();
        Single retval = 500;
        if( weight IsNot Nothing && weight.equalsIgnoreCase( "bold" ) )
        {
            retval = 900;
        }
        else if( weight IsNot Nothing && weight.equalsIgnoreCase( "light" ) )
        {
            retval = 100;
        }
        return retval;
    }

    /**
     * Set the weight of the font.
     *
     * @param fontWeight The new weight of the font.
     */
    public void setFontWeight( Single fontWeight )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * A string representing the preferred font stretch.
     *
     * @return The font stretch.
     */
    public String getFontStretch()
    {
        return null;
    }

    /**
     * This will set the font stretch.
     *
     * @param fontStretch The font stretch
     */
    public void setFontStretch( String fontStretch )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the font flags.
     *
     * @return The font flags.
     */
    public int getFlags()
    {
        //I believe that the only flag that AFM supports is the is fixed pitch
        return afm.isFixedPitch() ? 1 : 0;
    }

    /**
     * This will set the font flags.
     *
     * @param flags The new font flags.
     */
    public void setFlags( int flags )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the fonts bouding box.
     *
     * @return The fonts bouding box.
     */
    public PDRectangle getFontBoundingBox()
    {
        BoundingBox box = afm.getFontBBox();
        PDRectangle retval = null;
        if( box IsNot Nothing )
        {
            retval = new PDRectangle( box );
        }
        return retval;
    }

    /**
     * Set the fonts bounding box.
     *
     * @param rect The new bouding box.
     */
    public void setFontBoundingBox( PDRectangle rect )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the italic angle for the font.
     *
     * @return The italic angle.
     */
    public Single getItalicAngle()
    {
        return afm.getItalicAngle();
    }

    /**
     * This will set the italic angle for the font.
     *
     * @param angle The new italic angle for the font.
     */
    public void setItalicAngle( Single angle )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the ascent for the font.
     *
     * @return The ascent.
     */
    public Single getAscent()
    {
        return afm.getAscender();
    }

    /**
     * This will set the ascent for the font.
     *
     * @param ascent The new ascent for the font.
     */
    public void setAscent( Single ascent )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the descent for the font.
     *
     * @return The descent.
     */
    public Single getDescent()
    {
        return afm.getDescender();
    }

    /**
     * This will set the descent for the font.
     *
     * @param descent The new descent for the font.
     */
    public void setDescent( Single descent )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the leading for the font.
     *
     * @return The leading.
     */
    public Single getLeading()
    {
        //AFM does not support setting the leading so we will just ignore it.
        return 0f;
    }

    /**
     * This will set the leading for the font.
     *
     * @param leading The new leading for the font.
     */
    public void setLeading( Single leading )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the CapHeight for the font.
     *
     * @return The cap height.
     */
    public Single getCapHeight()
    {
        return afm.getCapHeight();
    }

    /**
     * This will set the cap height for the font.
     *
     * @param capHeight The new cap height for the font.
     */
    public void setCapHeight( Single capHeight )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the x height for the font.
     *
     * @return The x height.
     */
    public Single getXHeight()
    {
        return afm.getXHeight();
    }

    /**
     * This will set the x height for the font.
     *
     * @param xHeight The new x height for the font.
     */
    public void setXHeight( Single xHeight )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the stemV for the font.
     *
     * @return The stem v value.
     */
    public Single getStemV()
    {
        //afm does not have a stem v
        return 0;
    }

    /**
     * This will set the stem V for the font.
     *
     * @param stemV The new stem v for the font.
     */
    public void setStemV( Single stemV )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the stemH for the font.
     *
     * @return The stem h value.
     */
    public Single getStemH()
    {
        //afm does not have a stem h
        return 0;
    }

    /**
     * This will set the stem H for the font.
     *
     * @param stemH The new stem h for the font.
     */
    public void setStemH( Single stemH )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the average width for the font.
     *
     * @return The average width value.
     *
     * @throws IOException If there is an error calculating the average width.
     */
    public Single getAverageWidth() throws IOException
    {
        return afm.getAverageCharacterWidth();
    }

    /**
     * This will set the average width for the font.
     *
     * @param averageWidth The new average width for the font.
     */
    public void setAverageWidth( Single averageWidth )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the max width for the font.
     *
     * @return The max width value.
     */
    public Single getMaxWidth()
    {
        //afm does not support max width;
        return 0;
    }

    /**
     * This will set the max width for the font.
     *
     * @param maxWidth The new max width for the font.
     */
    public void setMaxWidth( Single maxWidth )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the missing width for the font.
     *
     * @return The missing width value.
     */
    public Single getMissingWidth()
    {
        return 0;
    }

    /**
     * This will set the missing width for the font.
     *
     * @param missingWidth The new missing width for the font.
     */
    public void setMissingWidth( Single missingWidth )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }

    /**
     * This will get the character set for the font.
     *
     * @return The character set value.
     */
    public String getCharSet()
    {
        return afm.getCharacterSet();
    }

    /**
     * This will set the character set for the font.
     *
     * @param charSet The new character set for the font.
     */
    public void setCharacterSet( String charSet )
    {
        throw new UnsupportedOperationException( "The AFM Font descriptor is immutable" );
    }
}

File: Projects/DMDPDF/PDFBox/pdmodel/font/PDFontDescriptorAFM.vb
Imports System.IO
Imports System.Drawing
Imports FinSeA.Drawings
Imports FinSeA.org.apache.fontbox.afm
Imports FinSeA.org.apache.pdfbox.pdmodel.common
Imports FinSeA.org.apache.fontbox.util


Namespace org.apache.pdfbox.pdmodel.font

    '/**
    ' * This class represents the font descriptor when the font information
    ' * is coming from an AFM file.
    ' *
    ' * @author <a href="mailto:ben@benlitchfield.com">Ben Litchfield</a>
    ' * @version $Revision: 1.3 $
    ' */
    Public Class PDFontDescriptorAFM
        Inherits PDFontDescriptor

        Private afm As FontMetric

        '/**
        ' * Constructor.
        ' *
        ' * @param afmFile The AFM file.
        ' */
        Public Sub New(ByVal afmFile As FontMetric)
            afm = afmFile
        End Sub

        '/**
        ' * Get the font name.
        ' *
        ' * @return The name of the font.
        ' */
        Public Overrides Function getFontName() As String
            Return afm.getFontName()
        End Function

        '/**
        ' * This will set the font name.
        ' *
        ' * @param fontName The new name for the font.
        ' */
        Public Overrides Sub setFontName(ByVal fontName As String)
            Me.RaiseUnsupported()
        End Sub

        Private Sub RaiseUnsupported()
            Throw New NotSupportedException("The AFM Font descriptor is immutable")
        End Sub

        '/**
        ' * A string representing the preferred font family.
        ' *
        ' * @return The font family.
        ' */
        Public Overrides Function getFontFamily() As String
            Return afm.getFamilyName()
        End Function

        '/**
        ' * This will set the font family.
        ' *
        ' * @param fontFamily The font family.
        ' */
        Public Overrides Sub setFontFamily(ByVal fontFamily As String)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * The weight of the font.  According to the PDF spec "possible values are
        ' * 100, 200, 300, 400, 500, 600, 700, 800 or 900"  Where a higher number is
        ' * more weight and appears to be more bold.
        ' *
        ' * @return The font weight.
        ' */
        Public Overrides Function getFontWeight() As Single
            Dim weight As String = afm.getWeight()
            Dim retval As Single = 500
            If (weight <> "" AndAlso LCase(weight).Equals("bold")) Then
                retval = 900
            ElseIf (weight <> "" AndAlso LCase(weight).Equals("light")) Then
                retval = 100
            End If
            Return retval
        End Function

        '/**
        ' * Set the weight of the font.
        ' *
        ' * @param fontWeight The new weight of the font.
        ' */
        Public Overrides Sub setFontWeight(ByVal fontWeight As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * A string representing the preferred font stretch.
        ' *
        ' * @return The font stretch.
        ' */
        Public Overrides Function getFontStretch() As String
            Return ""
        End Function

        '/**
        ' * This will set the font stretch.
        ' *
        ' * @param fontStretch The font stretch
        ' */
        Public Overrides Sub setFontStretch(ByVal fontStretch As String)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the font flags.
        ' *
        ' * @return The font flags.
        ' */
        Public Overrides Function getFlags() As Integer
            'I believe that the only flag that AFM supports is the is fixed pitch
            Return IIf(afm.isFixedPitch(), 1, 0)
        End Function

        '/**
        ' * This will set the font flags.
        ' *
        ' * @param flags The new font flags.
        ' */
        Public Overrides Sub setFlags(ByVal flags As Integer)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the fonts bouding box.
        ' *
        ' * @return The fonts bouding box.
        ' */
        Public Overrides Function getFontBoundingBox() As PDRectangle
            Dim box As BoundingBox = afm.getFontBBox()
            Dim retval As PDRectangle = Nothing
            If (box IsNot Nothing) Then
                retval = New PDRectangle(box)
            End If
            Return retval
        End Function

        '/**
        ' * Set the fonts bounding box.
        ' *
        ' * @param rect The new bouding box.
        ' */
        Public Overrides Sub setFontBoundingBox(ByVal rect As PDRectangle)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the italic angle for the font.
        ' *
        ' * @return The italic angle.
        ' */
        Public Overrides Function getItalicAngle() As Single
            Return afm.getItalicAngle()
        End Function

        '/**
        ' * This will set the italic angle for the font.
        ' *
        ' * @param angle The new italic angle for the font.
        ' */
        Public Overrides Sub setItalicAngle(ByVal angle As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the ascent for the font.
        ' *
        ' * @return The ascent.
        ' */
        Public Overrides Function getAscent() As Single
            Return afm.getAscender()
        End Function

        '/**
        ' * This will set the ascent for the font.
        ' *
        ' * @param ascent The new ascent for the font.
        ' */
        Public Overrides Sub setAscent(ByVal ascent As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the descent for the font.
        ' *
        ' * @return The descent.
        ' */
        Public Overrides Function getDescent() As Single
            Return afm.getDescender()
        End Function

        '/**
        ' * This will set the descent for the font.
        ' *
        ' * @param descent The new descent for the font.
        ' */
        Public Overrides Sub setDescent(ByVal descent As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the leading for the font.
        ' *
        ' * @return The leading.
        ' */
        Public Overrides Function getLeading() As Single
            'AFM does not support setting the leading so we will just ignore it.
            Return 0.0F
        End Function

        '/**
        ' * This will set the leading for the font.
        ' *
        ' * @param leading The new leading for the font.
        ' */
        Public Overrides Sub setLeading(ByVal leading As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the CapHeight for the font.
        ' *
        ' * @return The cap height.
        ' */
        Public Overrides Function getCapHeight() As Single
            Return afm.getCapHeight()
        End Function

        '/**
        ' * This will set the cap height for the font.
        ' *
        ' * @param capHeight The new cap height for the font.
        ' */
        Public Overrides Sub setCapHeight(ByVal capHeight As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the x height for the font.
        ' *
        ' * @return The x height.
        ' */
        Public Overrides Function getXHeight() As Single
            Return afm.getXHeight()
        End Function

        '/**
        ' * This will set the x height for the font.
        ' *
        ' * @param xHeight The new x height for the font.
        ' */
        Public Overrides Sub setXHeight(ByVal xHeight As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the stemV for the font.
        ' *
        ' * @return The stem v value.
        ' */
        Public Overrides Function getStemV() As Single
            'afm does not have a stem v
            Return 0
        End Function

        '/**
        ' * This will set the stem V for the font.
        ' *
        ' * @param stemV The new stem v for the font.
        ' */
        Public Overrides Sub setStemV(ByVal stemV As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the stemH for the font.
        ' *
        ' * @return The stem h value.
        ' */
        Public Overrides Function getStemH() As Single
            'afm does not have a stem h
            Return 0
        End Function

        '/**
        ' * This will set the stem H for the font.
        ' *
        ' * @param stemH The new stem h for the font.
        ' */
        Public Overrides Sub setStemH(ByVal stemH As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the average width for the font.
        ' *
        ' * @return The average width value.
        ' *
        ' * @throws IOException If there is an error calculating the average width.
        ' */
        Public Overrides Function getAverageWidth() As Single 'throws IOException
            Return afm.getAverageCharacterWidth()
        End Function

        '/**
        ' * This will set the average width for the font.
        ' *
        ' * @param averageWidth The new average width for the font.
        ' */
        Public Overrides Sub setAverageWidth(ByVal averageWidth As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the max width for the font.
        ' *
        ' * @return The max width value.
        ' */
        Public Overrides Function getMaxWidth() As Single
            'afm does not support max width;
            Return 0
        End Function

        '/**
        ' * This will set the max width for the font.
        ' *
        ' * @param maxWidth The new max width for the font.
        ' */
        Public Overrides Sub setMaxWidth(ByVal maxWidth As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the missing width for the font.
        ' *
        ' * @return The missing width value.
        ' */
        Public Overrides Function getMissingWidth() As Single
            Return 0
        End Function

        '/**
        ' * This will set the missing width for the font.
        ' *
        ' * @param missingWidth The new missing width for the font.
        ' */
        Public Overrides Sub setMissingWidth(ByVal missingWidth As Single)
            Me.RaiseUnsupported()
        End Sub

        '/**
        ' * This will get the character set for the font.
        ' *
        ' * @return The character set value.
        ' */
        Public Overrides Function getCharSet() As String
            Return afm.getCharacterSet()
        End Function

        '/**
        ' * This will set the character set for the font.
        ' *
        ' * @param charSet The new character set for the font.
        ' */
        Public Overrides Sub setCharacterSet(ByVal charSet As String)
            Me.RaiseUnsupported()
        End Sub


    End Class

End Namespace
